File: lib/quasifind/search_stubs.c
/* search_stubs.c - Mmap based regex search with optional SIMD acceleration */

#include <sys/mman.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <regex.h>
#include <string.h>
#include <errno.h>

#include <caml/mlvalues.h>
#include <caml/memory.h>
#include <caml/alloc.h>
#include <caml/fail.h>
#include <caml/unixsupport.h>
#include <caml/signals.h>

/* ═══════════════════════════════════════════════════════════════════════════
   ARM NEON SIMD-accelerated memmem
   ═══════════════════════════════════════════════════════════════════════════ */

#if defined(__ARM_NEON) || defined(__ARM_NEON__)
#include <arm_neon.h>

/*
   neon_extract_mask: Convert a NEON comparison result (0x00/0xFF per byte)
   into a 16-bit mask where each bit represents whether the corresponding
   byte matched.
*/
static inline uint16_t neon_extract_mask(uint8x16_t v)
{
    /* Shift each byte to isolate the high bit, then narrow and combine */
    static const uint8x16_t shift = {0, 1, 2, 3, 4, 5, 6, 7, 0, 1, 2, 3, 4, 5, 6, 7};
    uint8x16_t masked = vshrq_n_u8(v, 7); /* 0x00 or 0x01 per lane */
    uint8x16_t shifted = vshlq_u8(masked, vreinterpretq_s8_u8(shift));
    /* Sum the low 8 and high 8 lanes separately */
    uint8_t lo = vaddv_u8(vget_low_u8(shifted));
    /* vget_high returns uint8x8_t, use vaddv_u8 (not vaddvq_u8) */
    uint8x8_t hi_half = vget_high_u8(shifted);
    uint8_t hi = vaddv_u8(hi_half);
    return (uint16_t)lo | ((uint16_t)hi << 8);
}

/*
   simd_memmem_neon: SIMD-accelerated string search using ARM NEON.

   Algorithm:
   1. Broadcast needle[0] and needle[nlen-1] to all 16 NEON lanes
   2. For each 16-byte aligned chunk of haystack:
      a. Compare chunk[i] == needle[0] AND chunk[i+nlen-1] == needle[nlen-1]
      b. Extract a bitmask of matching positions
      c. For each match, verify with memcmp

   This "first+last" trick dramatically reduces false positives compared
   to single-byte scanning, especially for longer needles.
*/
static void *simd_memmem_neon(const void *haystack, size_t hlen,
                              const void *needle, size_t nlen)
{
    if (nlen == 0)
        return (void *)haystack;
    if (nlen > hlen)
        return NULL;

    const uint8_t *h = (const uint8_t *)haystack;
    const uint8_t *n = (const uint8_t *)needle;
    size_t end = hlen - nlen;

    /* Single byte needle: just use memchr */
    if (nlen == 1)
    {
        return memchr(haystack, n[0], hlen);
    }

    /* Broadcast first and last bytes of needle to all 16 lanes */
    uint8x16_t first = vdupq_n_u8(n[0]);
    uint8x16_t last = vdupq_n_u8(n[nlen - 1]);

    size_t i = 0;

    /* Process 16 bytes at a time */
    for (; i + 15 + nlen - 1 <= hlen; i += 16)
    {
        /* Load 16 bytes starting at position i (first byte candidates) */
        uint8x16_t block_first = vld1q_u8(h + i);
        /* Load 16 bytes starting at position i+nlen-1 (last byte candidates) */
        uint8x16_t block_last = vld1q_u8(h + i + nlen - 1);

        /* Compare: which positions have matching first AND last byte? */
        uint8x16_t eq_first = vceqq_u8(block_first, first);
        uint8x16_t eq_last = vceqq_u8(block_last, last);
        uint8x16_t eq = vandq_u8(eq_first, eq_last);

        /* Quick check: any matches at all? */
        if (vmaxvq_u8(eq) == 0)
            continue;

        /* Extract bitmask and check each candidate */
        uint16_t mask = neon_extract_mask(eq);
        while (mask != 0)
        {
            int bit = __builtin_ctz(mask);
            size_t pos = i + bit;
            if (pos <= end && memcmp(h + pos, n, nlen) == 0)
            {
                return (void *)(h + pos);
            }
            mask &= mask - 1; /* Clear lowest set bit */
        }
    }

    /* Handle remaining bytes with scalar fallback */
    for (; i <= end; i++)
    {
        if (h[i] == n[0] && memcmp(h + i, n, nlen) == 0)
        {
            return (void *)(h + i);
        }
    }

    return NULL;
}
/* ═══════════════════════════════════════════════════════════════════════════
   Lightweight Aho-Corasick / KMP + BMH + NEON hybrid search

   Total footprint: ~520 bytes (vs 263KB for full AC goto table)
   - Pattern pointer + length
   - KMP failure table (max 256 ints)
   - BMH skip table (256 ints)
   ═══════════════════════════════════════════════════════════════════════════ */

#define AC_MAX_PATTERN 256

typedef struct
{
    const uint8_t *pattern;
    size_t plen;
    int fail[AC_MAX_PATTERN]; /* KMP failure table */
    int skip[256];            /* BMH skip table */
} ac_lite_t;

static void ac_lite_build(ac_lite_t *ac, const uint8_t *pattern, size_t plen)
{
    ac->pattern = pattern;
    ac->plen = plen;

    /* KMP failure table */
    ac->fail[0] = -1;
    if (plen > 1)
    {
        ac->fail[1] = 0;
        for (size_t i = 2; i < plen && i < AC_MAX_PATTERN; i++)
        {
            int j = ac->fail[i - 1];
            while (j >= 0 && pattern[(size_t)j] != pattern[i - 1])
            {
                j = ac->fail[(size_t)j];
            }
            ac->fail[i] = j + 1;
        }
    }

    /* BMH skip table */
    for (int c = 0; c < 256; c++)
    {
        ac->skip[c] = (int)plen;
    }
    for (size_t i = 0; i + 1 < plen; i++)
    {
        ac->skip[pattern[i]] = (int)(plen - 1 - i);
    }
}

/*
   ac_lite_search_neon: NEON first+last trick + BMH scalar tail.
   Pattern bytes accessed directly (no goto table indirection).
*/
static void *ac_lite_search_neon(const ac_lite_t *ac,
                                 const void *haystack, size_t hlen)
{
    const uint8_t *h = (const uint8_t *)haystack;
    const uint8_t *p = ac->pattern;
    const size_t plen = ac->plen;

    if (plen == 0)
        return (void *)haystack;
    if (plen > hlen)
        return NULL;
    if (plen <= 3)
        return NULL; /* Signal: use simd_memmem_neon */

    const uint8_t first_byte = p[0];
    const uint8_t last_byte = p[plen - 1];
    const size_t end = hlen - plen;

    uint8x16_t vfirst = vdupq_n_u8(first_byte);
    uint8x16_t vlast = vdupq_n_u8(last_byte);

    size_t i = 0;

    /* NEON: check first+last byte simultaneously across 16 positions */
    for (; i + 15 + plen - 1 <= hlen; i += 16)
    {
        uint8x16_t eq_first = vceqq_u8(vld1q_u8(h + i), vfirst);
        uint8x16_t eq_last = vceqq_u8(vld1q_u8(h + i + plen - 1), vlast);
        uint8x16_t eq = vandq_u8(eq_first, eq_last);

        if (vmaxvq_u8(eq) == 0)
            continue;

        uint16_t mask = neon_extract_mask(eq);
        while (mask != 0)
        {
            int bit = __builtin_ctz(mask);
            size_t pos = i + (size_t)bit;
            if (pos <= end && memcmp(h + pos + 1, p + 1, plen - 2) == 0)
            {
                return (void *)(h + pos);
            }
            mask &= mask - 1;
        }
    }

    /* Scalar tail with BMH skip */
    size_t j = i + plen - 1;
    while (j < hlen)
    {
        if (h[j] != last_byte)
        {
            j += (size_t)ac->skip[h[j]];
            continue;
        }
        size_t start = j - (plen - 1);
        if (h[start] == first_byte && memcmp(h + start + 1, p + 1, plen - 2) == 0)
        {
            return (void *)(h + start);
        }
        j++;
    }

    return NULL;
}

#endif /* __ARM_NEON */

/*
   caml_search_regex(v_path, v_pattern)

   Returns:
     1 (true)  if pattern matches content
     0 (false) if pattern does not match
    -1 (error) if something went wrong (e.g. regex compilation failed, mmap failed)
               allowing fallback to OCaml implementation.
*/
CAMLprim value caml_search_regex(value v_path, value v_pattern)
{
    CAMLparam2(v_path, v_pattern);

#ifndef REG_STARTEND
    /* Without REG_STARTEND, we cannot safely use regexec on mmap'd data
       because the mapped region is not guaranteed to be null-terminated.
       We fail explicitly here to prevent out-of-bounds reads. */
    caml_failwith("Regex search on mmap requires REG_STARTEND support on this OS.");
#else
    const char *path = String_val(v_path);
    const char *pattern = String_val(v_pattern);
    int ret = 0;

    /* 1. Compile Regex */
    regex_t regex;
    /* Use REG_EXTENDED for modern syntax, REG_NOSUB for speed (we only care if it matches) */
    /* Note: macOS/BSD regex might differ slightly from GNU, hence fallback strategy is crucial. */
    if (regcomp(&regex, pattern, REG_EXTENDED | REG_NOSUB | REG_NEWLINE) != 0)
    {
        /* Regex compilation failed - likely unsupported syntax */
        CAMLreturn(Val_int(-1));
    }

    /* 2. Open File */
    int fd = open(path, O_RDONLY);
    if (fd == -1)
    {
        regfree(&regex);
        /* If file cannot be opened, we can't search it. Return -1 to let OCaml implementation handle it (or fail there). */
        CAMLreturn(Val_int(-1));
    }

    /* 3. Get file size */
    struct stat st;
    if (fstat(fd, &st) == -1)
    {
        close(fd);
        regfree(&regex);
        CAMLreturn(Val_int(-1));
    }

    if (st.st_size == 0)
    {
        /* Empty file matches nothing usually, checks against empty string?
           Let's return 0 (no match) or check against empty buffer. */
        /* regexec on empty string: */
        int match = regexec(&regex, "", 0, NULL, 0);
        close(fd);
        regfree(&regex);
        CAMLreturn(Val_int(match == 0 ? 1 : 0));
    }

    /* 4. Mmap */
    /* MAP_PRIVATE ensures we don't write back changes (though we are PROT_READ anyway) */
    void *addr = mmap(NULL, st.st_size, PROT_READ, MAP_PRIVATE, fd, 0);
    if (addr == MAP_FAILED)
    {
        close(fd);
        regfree(&regex);
        CAMLreturn(Val_int(-1));
    }

    /* 5. Execute Regex */
    /* We use REG_STARTEND to safely bound the search within the mmap'd size */
    regmatch_t pmatch[1];
    pmatch[0].rm_so = 0;
    pmatch[0].rm_eo = st.st_size;

    /* Release OCaml runtime lock while executing regex to allow other fibers/domains to proceed */
    caml_enter_blocking_section();
    int rc = regexec(&regex, (const char *)addr, 0, pmatch, REG_STARTEND);
    caml_leave_blocking_section();

    if (rc == 0)
    {
        ret = 1; /* Match */
    }
    else
    {
        ret = 0; /* No match (REG_NOMATCH) */
    }

    /* 6. Cleanup */
    munmap(addr, st.st_size);
    close(fd);
    regfree(&regex);

    CAMLreturn(Val_int(ret));
#endif
}

/*
   caml_search_memmem(v_path, v_needle)

   Fast literal string search using mmap + memmem.
   memmem on macOS/glibc uses SIMD-optimized routines internally.

   Returns:
     1 (true)  if needle is found in file content
     0 (false) if needle is not found
    -1 (error) if something went wrong
*/
CAMLprim value caml_search_memmem(value v_path, value v_needle)
{
    CAMLparam2(v_path, v_needle);
    const char *path = String_val(v_path);
    const char *needle = String_val(v_needle);
    mlsize_t needle_len = caml_string_length(v_needle);

    if (needle_len == 0)
    {
        /* Empty needle matches everything */
        CAMLreturn(Val_int(1));
    }

    /* 1. Open File */
    int fd = open(path, O_RDONLY);
    if (fd == -1)
    {
        CAMLreturn(Val_int(-1));
    }

    /* 2. Get file size */
    struct stat st;
    if (fstat(fd, &st) == -1 || st.st_size == 0)
    {
        close(fd);
        CAMLreturn(Val_int(st.st_size == 0 ? 0 : -1));
    }

    /* Skip files larger than 256MB to avoid excessive memory mapping */
    if (st.st_size > 256 * 1024 * 1024)
    {
        close(fd);
        CAMLreturn(Val_int(-1));
    }

    /* 3. Mmap */
    void *addr = mmap(NULL, st.st_size, PROT_READ, MAP_PRIVATE, fd, 0);
    if (addr == MAP_FAILED)
    {
        close(fd);
        CAMLreturn(Val_int(-1));
    }

    /* 4. Search - use SIMD-accelerated version on ARM, libc memmem elsewhere */
    caml_enter_blocking_section();
#if defined(__ARM_NEON) || defined(__ARM_NEON__)
    void *found = NULL;
    /* For patterns >= 4 bytes, try AC/KMP + NEON pre-filter */
    if (needle_len >= 4 && needle_len <= AC_MAX_PATTERN)
    {
        ac_lite_t ac;
        ac_lite_build(&ac, (const uint8_t *)needle, needle_len);
        found = ac_lite_search_neon(&ac, addr, st.st_size);
    }
    /* Fallback: NEON memmem (first+last trick) for short patterns or AC delegation */
    if (found == NULL && needle_len < 4)
    {
        found = simd_memmem_neon(addr, st.st_size, needle, needle_len);
    }
    /* Final fallback: simple NEON memmem if AC found nothing */
    if (found == NULL && needle_len >= 4)
    {
        found = simd_memmem_neon(addr, st.st_size, needle, needle_len);
    }
#else
    void *found = memmem(addr, st.st_size, needle, needle_len);
#endif
    caml_leave_blocking_section();
    
    int ret = (found != NULL) ? 1 : 0;

    /* 5. Cleanup */
    munmap(addr, st.st_size);
    close(fd);

    CAMLreturn(Val_int(ret));
}

File: lib/quasifind/search_stubs.zig
const std = @import("std");
const builtin = @import("builtin");
const c = @cImport({
    @cInclude("sys/mman.h");
    @cInclude("sys/stat.h");
    @cInclude("fcntl.h");
    @cInclude("unistd.h");
    @cInclude("regex.h");
    @cInclude("string.h");
    @cInclude("caml/mlvalues.h");
    @cInclude("caml/memory.h");
    @cInclude("caml/alloc.h");
    @cInclude("caml/fail.h");
    @cInclude("caml/unixsupport.h");
    @cInclude("caml/signals.h");
});

inline fn String_val(v: c.value) [*c]const u8 {
    return @as([*c]const u8, @ptrFromInt(@as(usize, @bitCast(v))));
}

inline fn Int_val(v: c.value) c_int {
    return @as(c_int, @intCast(@as(isize, @bitCast(v)) >> 1));
}

inline fn Val_int(i: c_int) c.value {
    return @as(c.value, @bitCast((@as(isize, i) << 1) | 1));
}

// caml_string_length is a complex macro in OCaml, so we read the header safely in Zig.
inline fn caml_string_length(v: c.value) usize {
    // String length is stored in the header word before the string data.
    // OCaml string block: [header (8 bytes)] [chars...] [\0]
    // The last byte of the block contains padding info.
    // Since Zig is just linking against OCaml, we can rely on OCaml's caml_string_length macro if correctly imported,
    // but if it fails to compile, we implement it manually: Size of block - padding - 1.
    // Actually, `caml_string_length` function exists in OCaml 5 C API natively for simplicity, but let's see.
    return c.caml_string_length(v);
}

// SIMD memmem implementation
inline fn neon_extract_mask(v: @Vector(16, u8)) u16 {
    // We select bits per lane and sum.
    const bit_select = @Vector(16, u8){
        1, 2, 4, 8, 16, 32, 64, 128,
        1, 2, 4, 8, 16, 32, 64, 128,
    };
    const selected = v & bit_select;

    // Sum lower 8 and upper 8 bytes.
    const low_sum = @reduce(.Add, @as(@Vector(8, u8), @shuffle(u8, selected, undefined, @Vector(8, i32){ 0, 1, 2, 3, 4, 5, 6, 7 })));
    const high_sum = @reduce(.Add, @as(@Vector(8, u8), @shuffle(u8, selected, undefined, @Vector(8, i32){ 8, 9, 10, 11, 12, 13, 14, 15 })));

    return @as(u16, low_sum) | (@as(u16, high_sum) << 8);
}

fn simd_memmem_neon(haystack: [*]const u8, hlen: usize, needle: [*]const u8, nlen: usize) ?[*]const u8 {
    if (nlen == 0) return haystack;
    if (nlen > hlen) return null;

    if (nlen == 1) {
        if (c.memchr(haystack, needle[0], hlen)) |found| {
            return @as([*]const u8, @ptrCast(found));
        }
        return null;
    }

    const first_byte = needle[0];
    const last_byte = needle[nlen - 1];
    const end = hlen - nlen;

    const vfirst: @Vector(16, u8) = @splat(first_byte);
    const vlast: @Vector(16, u8) = @splat(last_byte);

    var i: usize = 0;

    // Unroll by 2 (32 bytes per loop iteration) with memory prefetch
    while (i + 31 + nlen - 1 <= hlen) : (i += 32) {
        // Prefetch 256 bytes ahead into L3 cache
        @prefetch(haystack + i + 256, .{ .rw = .read, .locality = 3, .cache = .data });

        comptime var u: usize = 0;
        inline while (u < 2) : (u += 1) {
            const offset = i + (u * 16);

            const block_first: @Vector(16, u8) = haystack[offset .. offset + 16][0..16].*;
            const block_last: @Vector(16, u8) = haystack[offset + nlen - 1 .. offset + nlen - 1 + 16][0..16].*;

            const eq_first = block_first == vfirst;
            const eq_last = block_last == vlast;

            const eq_first_u8 = @select(u8, eq_first, @as(@Vector(16, u8), @splat(0xFF)), @as(@Vector(16, u8), @splat(0x00)));
            const eq_last_u8 = @select(u8, eq_last, @as(@Vector(16, u8), @splat(0xFF)), @as(@Vector(16, u8), @splat(0x00)));
            const eq = eq_first_u8 & eq_last_u8;

            const max_val = @reduce(.Max, eq);
            if (max_val > 0) {
                var mask = neon_extract_mask(eq);
                while (mask != 0) {
                    const bit = @ctz(mask);
                    const pos = offset + bit;
                    if (pos <= end and c.memcmp(haystack + pos, needle, nlen) == 0) {
                        return haystack + pos;
                    }
                    mask &= mask - 1;
                }
            }
        }
    }

    while (i <= end) : (i += 1) {
        if (haystack[i] == first_byte and c.memcmp(haystack + i, needle, nlen) == 0) {
            return haystack + i;
        }
    }

    return null;
}

const AC_MAX_PATTERN = 256;
const ac_lite_t = struct {
    pattern: [*]const u8,
    plen: usize,
    fail: [AC_MAX_PATTERN]i32,
    skip: [256]i32,
};

fn ac_lite_build(ac: *ac_lite_t, pattern: [*]const u8, plen: usize) void {
    ac.pattern = pattern;
    ac.plen = plen;

    ac.fail[0] = -1;
    if (plen > 1) {
        ac.fail[1] = 0;
        var i: usize = 2;
        while (i < plen and i < AC_MAX_PATTERN) : (i += 1) {
            var j = ac.fail[i - 1];
            while (j >= 0 and pattern[@intCast(j)] != pattern[i - 1]) {
                j = ac.fail[@intCast(j)];
            }
            ac.fail[i] = j + 1;
        }
    }

    for (&ac.skip) |*s| {
        s.* = @intCast(plen);
    }
    var i: usize = 0;
    while (i + 1 < plen) : (i += 1) {
        ac.skip[pattern[i]] = @intCast(plen - 1 - i);
    }
}

fn ac_lite_search_neon(ac: *const ac_lite_t, haystack: [*]const u8, hlen: usize) ?[*]const u8 {
    const p = ac.pattern;
    const plen = ac.plen;

    if (plen == 0) return haystack;
    if (plen > hlen) return null;

    // Explicit Contract:
    // This function returns null for patterns of length <= 3.
    // Short patterns lack the requisite entropy for Aho-Corasick skips, making the
    // overhead of this function unjustifiable compared to the SIMD memmem fallback.
    // The caller (e.g., caml_search_memmem) MUST provide an alternative fallback mechanism
    // if this function yields null.
    if (plen <= 3) return null;

    const first_byte = p[0];
    const last_byte = p[plen - 1];
    const end = hlen - plen;

    const vfirst: @Vector(16, u8) = @splat(first_byte);
    const vlast: @Vector(16, u8) = @splat(last_byte);

    var i: usize = 0;

    while (i + 15 + plen - 1 <= hlen) : (i += 16) {
        const block_first: @Vector(16, u8) = haystack[i .. i + 16][0..16].*;
        const block_last: @Vector(16, u8) = haystack[i + plen - 1 .. i + plen - 1 + 16][0..16].*;

        const eq_first = block_first == vfirst;
        const eq_last = block_last == vlast;

        const eq_first_u8 = @select(u8, eq_first, @as(@Vector(16, u8), @splat(0xFF)), @as(@Vector(16, u8), @splat(0x00)));
        const eq_last_u8 = @select(u8, eq_last, @as(@Vector(16, u8), @splat(0xFF)), @as(@Vector(16, u8), @splat(0x00)));
        const eq = eq_first_u8 & eq_last_u8;

        const max_val = @reduce(.Max, eq);
        if (max_val > 0) {
            var mask = neon_extract_mask(eq);
            while (mask != 0) {
                const bit = @ctz(mask);
                const pos = i + bit;
                if (pos <= end and c.memcmp(haystack + pos + 1, p + 1, plen - 2) == 0) {
                    return haystack + pos;
                }
                mask &= mask - 1;
            }
        }
    }

    var j: usize = i + plen - 1;
    while (j < hlen) {
        if (haystack[j] != last_byte) {
            const skip_val = ac.skip[haystack[j]];
            j += @intCast(skip_val);
            continue;
        }
        const start = j - (plen - 1);
        if (haystack[start] == first_byte and c.memcmp(haystack + start + 1, p + 1, plen - 2) == 0) {
            return haystack + start;
        }
        j += 1;
    }

    return null;
}

export fn caml_search_regex(v_path: c.value, v_pattern: c.value) c.value {
    const path = String_val(v_path);
    const pattern = String_val(v_pattern);
    var ret: c_int = 0;

    var regex: c.regex_t = undefined;
    if (c.regcomp(&regex, pattern, c.REG_EXTENDED | c.REG_NOSUB | c.REG_NEWLINE) != 0) {
        return Val_int(-1);
    }

    const fd = c.open(path, c.O_RDONLY);
    if (fd == -1) {
        c.regfree(&regex);
        return Val_int(-1);
    }

    var st: c.struct_stat = undefined;
    if (c.fstat(fd, &st) == -1) {
        _ = c.close(fd);
        c.regfree(&regex);
        return Val_int(-1);
    }

    if (st.st_size == 0) {
        const match = c.regexec(&regex, "", 0, null, 0);
        _ = c.close(fd);
        c.regfree(&regex);
        return Val_int(if (match == 0) 1 else 0);
    }

    const addr = c.mmap(null, @intCast(st.st_size), c.PROT_READ, c.MAP_PRIVATE, fd, 0);
    if (addr == c.MAP_FAILED) {
        _ = c.close(fd);
        c.regfree(&regex);
        return Val_int(-1);
    }

    var pmatch: [1]c.regmatch_t = undefined;
    pmatch[0].rm_so = 0;
    pmatch[0].rm_eo = st.st_size;

    c.caml_enter_blocking_section();
    // Use REG_STARTEND if available (macOS/BSD), otherwise fallback to null-terminated buffer (Linux)
    var rc: c_int = c.REG_NOMATCH;
    if (@hasDecl(c, "REG_STARTEND")) {
        rc = c.regexec(&regex, @ptrCast(addr), 0, &pmatch[0], c.REG_STARTEND);
    } else {
        const buf = c.malloc(@as(usize, @intCast(st.st_size)) + 1);
        if (buf != null) {
            const buf_ptr = @as([*c]u8, @ptrCast(buf));
            @memcpy(buf_ptr[0..@as(usize, @intCast(st.st_size))], @as([*]const u8, @ptrCast(addr))[0..@as(usize, @intCast(st.st_size))]);
            buf_ptr[@as(usize, @intCast(st.st_size))] = 0;
            rc = c.regexec(&regex, buf_ptr, 0, &pmatch[0], 0);
            c.free(buf);
        }
    }
    c.caml_leave_blocking_section();

    ret = if (rc == 0) 1 else 0;

    _ = c.munmap(addr, @intCast(st.st_size));
    _ = c.close(fd);
    c.regfree(&regex);

    return Val_int(ret);
}

export fn caml_search_memmem(v_path: c.value, v_needle: c.value) c.value {
    const path = String_val(v_path);
    const needle = String_val(v_needle);
    const needle_len = caml_string_length(v_needle);

    if (needle_len == 0) {
        return Val_int(1);
    }

    const fd = c.open(path, c.O_RDONLY);
    if (fd == -1) {
        return Val_int(-1);
    }

    var st: c.struct_stat = undefined;
    if (c.fstat(fd, &st) == -1) {
        _ = c.close(fd);
        return Val_int(-1);
    }
    if (st.st_size == 0) {
        _ = c.close(fd);
        return Val_int(0);
    }

    if (st.st_size > 256 * 1024 * 1024) {
        _ = c.close(fd);
        return Val_int(-1);
    }

    const addr = c.mmap(null, @intCast(st.st_size), c.PROT_READ, c.MAP_PRIVATE, fd, 0);
    if (addr == c.MAP_FAILED) {
        _ = c.close(fd);
        return Val_int(-1);
    }

    c.caml_enter_blocking_section();
    const haystack = @as([*]const u8, @ptrCast(addr));
    const hlen = @as(usize, @intCast(st.st_size));
    const n = @as([*]const u8, @ptrCast(needle));

    var found: ?[*]const u8 = null;

    if (needle_len >= 4 and needle_len <= AC_MAX_PATTERN) {
        var ac: ac_lite_t = undefined;
        ac_lite_build(&ac, n, needle_len);
        found = ac_lite_search_neon(&ac, haystack, hlen);
    }

    if (found == null) {
        found = simd_memmem_neon(haystack, hlen, n, needle_len);
    }

    c.caml_leave_blocking_section();

    const ret: c_int = if (found != null) 1 else 0;

    _ = c.munmap(addr, @intCast(st.st_size));
    _ = c.close(fd);

    return Val_int(ret);
}
